packages/yew-router-macro/src/routable_derive.rs

```diff
@@ -22,225 +26,457 @@ impl Parse for Routable {
         let data = match data {
             Data::Enum(data) => data,
             Data::Struct(s) => {
-                return Err(syn::Error::new(
-                    s.struct_token.span(),
+                return Err(syn::Error::new_spanned(
+                    s.struct_token,
                     "expected enum, found struct",
                 ))
             }
             Data::Union(u) => {
-                return Err(syn::Error::new(
-                    u.union_token.span(),
+                return Err(syn::Error::new_spanned(
+                    u.union_token,
                     "expected enum, found union",
                 ))
             }
         };
 
-        let (not_found_route, ats) = parse_variants_attributes(&data.variants)?;
+        let variants = parse_variants(&ident, &data.variants)?;
+
+        Ok(Self { ident, variants })
+    }
+}
+
+#[derive(Debug)]
+struct ParsedField {
+    name: String,
+    index: usize,
+    meta: Option<Meta>,
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+enum BindMode {
+    Rest,
+    Path,
+    QueryArg,
+    HashArg,
+}
 
-        Ok(Self {
-            ident,
-            variants: data.variants,
-            ats,
-            not_found_route,
+enum BindType {
+    String,
+    Rest,
+}
+
+impl BindType {
+    fn decoder(&self) -> TokenStream {
+        let hidden = hidden_module();
+        match self {
+            BindType::Rest => quote! { #hidden::Routable::from_route(args.take_route()).ok() },
+            BindType::String => quote! { args.pop_str() },
+        }
+    }
+    fn encoder(&self, arg: &Ident) -> TokenStream {
+        let hidden = hidden_module();
+        match self {
+            BindType::Rest => quote! { args.store_route(#hidden::Routable::to_route(#arg)); },
+            BindType::String => quote! { args.push_str(#arg); },
+        }
+    }
+}
+
+struct ConstMuncher {
+    muncher: Option<TokenStream>,
+    binds: Vec<(usize, BindType)>,
+}
+
+fn generate_path_muncher(fields: &[ParsedField], at: &ParsedAt) -> syn::Result<ConstMuncher> {
+    let hidden = hidden_module();
+    let mut binds = Vec::new();
+    for name in &at.names {
+        if let Some(field) = fields.iter().find(|field| &field.name == name) {
+            binds.push((field.index, BindType::String));
+        } else {
+            return Err(syn::Error::new(
+                at.span,
+                format_args!("No field named {:?}", name),
```

Comment:
Please don't make the first letter uppercase, for consistency. There are many cases with this issue so please fix all of them.

Reply:
👍